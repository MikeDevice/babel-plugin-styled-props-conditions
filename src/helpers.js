export const findMatchingBrace = (chunks = []) => {
  if (!chunks.length || !chunks[0] || chunks[0][0] !== '{') return;

  let bracesCounter = 0;
  let chunkIndex;
  let charIndex;

  chunks.every((chunk, _chunkIndex) => {
    chunkIndex = _chunkIndex;

    Array.from(chunk).every((char, _charIndex) => {
      charIndex = _charIndex;

      switch (char) {
        case '{':
          bracesCounter += 1;
          break;

        case '}':
          bracesCounter -= 1;
          break;

        default:
          break;
      }

      return bracesCounter;
    });

    return bracesCounter;
  });

  if (bracesCounter === 0) {
    return { chunkIndex, charIndex };
  }
};

const parseCondition = (str) => {
  const matchResult = str.match(/^[_a-z]\w*/i);

  // There is no need to check variable 'matchResult' to being truthy
  // because this function is a helper function that is called only from
  // 'findConditionalBlocks' helper function and that function guarantees
  // to pass here the correct 'str' paramether

  const [prop] = matchResult;
  const { input } = matchResult;

  return {
    prop,
    rest: input.slice(prop.length),
  };
};

export const findConditionalBlocks = (chunks = []) => {
  const conditionalBlocks = [];

  chunks.forEach((chunk, chunkIndex) => {
    const matchResults = Array.from(chunk.matchAll(/@if(\s+[_a-z][^{]*){/gi));

    matchResults.forEach((result) => {
      const [match, condition] = result;
      const { index, input } = result;
      const bodyStartIndex = index + match.length - 1;

      const subChunks = chunks.slice(chunkIndex);
      subChunks[0] = input.slice(bodyStartIndex);

      const matchingBrace = findMatchingBrace(subChunks);
      const parsedCondition = parseCondition(condition.trim());

      if (matchingBrace && parsedCondition) {
        conditionalBlocks.push({
          chunkStart: chunkIndex,
          chunkEnd: chunkIndex + matchingBrace.chunkIndex,

          blockStart: index,
          blockEnd: matchingBrace.chunkIndex === 0
            ? matchingBrace.charIndex + bodyStartIndex
            : matchingBrace.charIndex,

          condition: parsedCondition,
          bodyStart: bodyStartIndex,
        });
      }
    });
  });

  return conditionalBlocks;
};

export const createExpression = ({ prop, rest }, hint) =>
  `\${({${prop}}) => ${prop}${rest} && ${hint}\``;
