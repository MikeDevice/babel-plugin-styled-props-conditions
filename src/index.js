import generate from '@babel/generator';
import { parse } from '@babel/parser';
import { addNamed } from '@babel/helper-module-imports';
import { findConditionalBlocks, createExpression } from './helpers';

export default function ({ types: t }) {
  return {
    name: 'styled-props-conditions',

    visitor: {
      TaggedTemplateExpression(path) {
        const { quasis } = path.node.quasi;
        const quasisRaws = quasis.map(({ value }) => value.raw);
        const conditionalBlocks = findConditionalBlocks(quasisRaws);
        const replaceData = [];

        if (!conditionalBlocks.length) return;

        const cssHint = addNamed(path, 'css', 'styled-components');

        conditionalBlocks.forEach((conditionalBlock) => {
          const {
            blockStart, blockEnd, bodyStart, chunkStart, chunkEnd, condition,
          } = conditionalBlock;

          replaceData.push(
            {
              quasiIndex: chunkEnd,
              start: blockEnd,
              del: 0,
              str: '`',
            },
            {
              quasiIndex: chunkStart,
              start: blockStart,
              del: bodyStart - blockStart + 1,
              str: createExpression(condition, cssHint.name),
            },
          );
        });

        const sortedArray = replaceData.sort(
          (a, b) => b.quasiIndex - a.quasiIndex || b.start - a.start,
        );

        sortedArray.forEach(({ quasiIndex, start, del, str }) => {
          const arr = [...quasisRaws[quasiIndex]];

          arr.splice(start, del, str);

          quasisRaws[quasiIndex] = arr.join('');
        });

        quasisRaws.forEach((raw, index) => {
          path.node.quasi.quasis[index] = t.templateElement({ raw });
        });

        const { code } = generate(path.node);
        const { expression } = parse(code).program.body[0];

        path.replaceWith(expression);
      },
    },
  };
}
