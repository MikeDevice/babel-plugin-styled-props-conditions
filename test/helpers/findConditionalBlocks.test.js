import { findConditionalBlocks } from '../../src/helpers';

const itHasNoBlocks = (message, chunks) => {
  it(message, () => {
    const result = findConditionalBlocks(chunks);
    expect(result).toEqual([]);
  });
};

const itHasBlocks = (message, chunks, expectedResult) => {
  it(message, () => {
    const result = findConditionalBlocks(chunks);
    expect(result).toEqual(expectedResult);
  });
};

describe('with no conditional blocks', () => {
  describe('no conditions', () => {
    itHasNoBlocks('no chunks');
    itHasNoBlocks('empty chunks', []);
    itHasNoBlocks('single chunk', ['color: red']);
    itHasNoBlocks('multiple chunks', ['a', '{ b }', '{ if a }']);
  });
  describe('incorrect conditions', () => {
    describe('single chunk', () => {
      itHasNoBlocks('no condition body', ['@if']);
      itHasNoBlocks('no brace', ['@if a']);
      itHasNoBlocks('no matching brace', ['@if a {']);
      itHasNoBlocks('incorrect keyword', ['@ifs a {}']);
      itHasNoBlocks('without identifier', ['@if + {}']);
      itHasNoBlocks('operator before identifier', ['@if + a {}']);
    });
    describe('multiple chunks', () => {
      itHasNoBlocks('keyword in multiple chunks', ['@i', 'f']);
      itHasNoBlocks('no brace', ['@if', 'a']);
      itHasNoBlocks('no matching brace', ['@if', ' a ', '{']);
      itHasNoBlocks('incorrect keyword', ['@ifs', ' a ', '{']);
      itHasNoBlocks('without identifier', ['@if', ' + ', '{}']);
      itHasNoBlocks('operator before identifier', ['@if', ' + a ', '{}']);
      itHasNoBlocks('correct but in multiple chunks', ['@if ', ' a ', '{}']);
    });
  });
});

describe('with conditional blocks', () => {
  describe('single chunk', () => {
    describe('without body', () => {
      itHasBlocks('simple prop', ['@if a {}'], [
        {
          chunkStart: 0,
          chunkEnd: 0,
          blockStart: 0,
          blockEnd: 7,
          bodyStart: 6,
          condition: { prop: 'a', rest: '' },
        },
      ]);
      itHasBlocks('prop with operations', ['@if a + 1 {}'], [
        {
          chunkStart: 0,
          chunkEnd: 0,
          blockStart: 0,
          blockEnd: 11,
          bodyStart: 10,
          condition: { prop: 'a', rest: ' + 1' },
        },
      ]);
      itHasBlocks('complex condition', ['@if b === [1, 2].join(" ") {}'], [
        {
          chunkStart: 0,
          chunkEnd: 0,
          blockStart: 0,
          blockEnd: 28,
          bodyStart: 27,
          condition: { prop: 'b', rest: ' === [1, 2].join(" ")' },
        },
      ]);
      itHasBlocks(
        'multiple blocks',
        [
          `
            @if a {}
            @if a === b {}
          `,
        ],
        [
          {
            chunkStart: 0,
            chunkEnd: 0,
            blockStart: 13,
            blockEnd: 20,
            bodyStart: 19,
            condition: { prop: 'a', rest: '' },
          },
          {
            chunkStart: 0,
            chunkEnd: 0,
            blockStart: 34,
            blockEnd: 47,
            bodyStart: 46,
            condition: { prop: 'a', rest: ' === b' },
          },
        ],
      );
    });
    describe('with body', () => {
      itHasBlocks(
        'simple prop',
        [
          '@if a { color: red; }',
        ],
        [
          {
            chunkStart: 0,
            chunkEnd: 0,
            blockStart: 0,
            blockEnd: 20,
            bodyStart: 6,
            condition: { prop: 'a', rest: '' },
          },
        ],
      );
      itHasBlocks(
        'prop with operations',
        [
          '@if variable !== a + b * 2  { background: white; }',
        ],
        [
          {
            chunkStart: 0,
            chunkEnd: 0,
            blockStart: 0,
            blockEnd: 49,
            bodyStart: 28,
            condition: { prop: 'variable', rest: ' !== a + b * 2' },
          },
        ],
      );
      itHasBlocks(
        'complex condition',
        [
          '@if b == func("ab") + func(["a", "b"].join("")) { color: white; }',
        ],
        [
          {
            chunkStart: 0,
            chunkEnd: 0,
            blockStart: 0,
            blockEnd: 64,
            bodyStart: 48,
            condition: { prop: 'b', rest: ' == func("ab") + func(["a", "b"].join(""))' },
          },
        ],
      );
      itHasBlocks(
        'nested blocks',
        [
          `@if a {
            color: red;

            @if b === "text" {
              color: green;
            }
          }`,
        ],
        [
          {
            chunkStart: 0,
            chunkEnd: 0,
            blockStart: 0,
            blockEnd: 116,
            bodyStart: 6,
            condition: { prop: 'a', rest: '' },
          },
          {
            chunkStart: 0,
            chunkEnd: 0,
            blockStart: 45,
            blockEnd: 104,
            bodyStart: 62,
            condition: { prop: 'b', rest: ' === "text"' },
          },
        ],
      );
    });
  });
});
