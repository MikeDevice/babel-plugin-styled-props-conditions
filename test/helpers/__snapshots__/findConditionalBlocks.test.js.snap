// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with conditional blocks multiple chunks complex condition 1`] = `
Array [
  Object {
    "blockEnd": 0,
    "blockStart": 0,
    "bodyStart": 24,
    "chunkEnd": 3,
    "chunkStart": 0,
    "condition": Object {
      "prop": "variable",
      "rest": " === \\"text\\"",
    },
  },
]
`;

exports[`with conditional blocks multiple chunks multiple blocks 1`] = `
Array [
  Object {
    "blockEnd": 0,
    "blockStart": 0,
    "bodyStart": 6,
    "chunkEnd": 2,
    "chunkStart": 0,
    "condition": Object {
      "prop": "b",
      "rest": "",
    },
  },
  Object {
    "blockEnd": 0,
    "blockStart": 0,
    "bodyStart": 23,
    "chunkEnd": 9,
    "chunkStart": 6,
    "condition": Object {
      "prop": "type",
      "rest": " === \\"default\\"",
    },
  },
]
`;

exports[`with conditional blocks multiple chunks nested blocks 1`] = `
Array [
  Object {
    "blockEnd": 0,
    "blockStart": 0,
    "bodyStart": 6,
    "chunkEnd": 6,
    "chunkStart": 0,
    "condition": Object {
      "prop": "b",
      "rest": "",
    },
  },
  Object {
    "blockEnd": 27,
    "blockStart": 2,
    "bodyStart": 14,
    "chunkEnd": 4,
    "chunkStart": 2,
    "condition": Object {
      "prop": "a",
      "rest": " === b",
    },
  },
  Object {
    "blockEnd": 24,
    "blockStart": 4,
    "bodyStart": 10,
    "chunkEnd": 3,
    "chunkStart": 3,
    "condition": Object {
      "prop": "c",
      "rest": "",
    },
  },
]
`;

exports[`with conditional blocks multiple chunks simple prop 1`] = `
Array [
  Object {
    "blockEnd": 0,
    "blockStart": 0,
    "bodyStart": 6,
    "chunkEnd": 2,
    "chunkStart": 0,
    "condition": Object {
      "prop": "a",
      "rest": "",
    },
  },
]
`;

exports[`with conditional blocks single chunk with body complex condition 1`] = `
Array [
  Object {
    "blockEnd": 91,
    "blockStart": 9,
    "bodyStart": 57,
    "chunkEnd": 0,
    "chunkStart": 0,
    "condition": Object {
      "prop": "b",
      "rest": " == func(\\"ab\\") + func([\\"a\\", \\"b\\"].join(\\"\\"))",
    },
  },
]
`;

exports[`with conditional blocks single chunk with body nested blocks 1`] = `
Array [
  Object {
    "blockEnd": 115,
    "blockStart": 9,
    "bodyStart": 15,
    "chunkEnd": 0,
    "chunkStart": 0,
    "condition": Object {
      "prop": "a",
      "rest": "",
    },
  },
  Object {
    "blockEnd": 105,
    "blockStart": 50,
    "bodyStart": 67,
    "chunkEnd": 0,
    "chunkStart": 0,
    "condition": Object {
      "prop": "b",
      "rest": " === \\"text\\"",
    },
  },
]
`;

exports[`with conditional blocks single chunk with body prop with operations 1`] = `
Array [
  Object {
    "blockEnd": 76,
    "blockStart": 9,
    "bodyStart": 37,
    "chunkEnd": 0,
    "chunkStart": 0,
    "condition": Object {
      "prop": "variable",
      "rest": " !== a + b * 2",
    },
  },
]
`;

exports[`with conditional blocks single chunk with body simple prop 1`] = `
Array [
  Object {
    "blockEnd": 20,
    "blockStart": 0,
    "bodyStart": 6,
    "chunkEnd": 0,
    "chunkStart": 0,
    "condition": Object {
      "prop": "a",
      "rest": "",
    },
  },
]
`;

exports[`with conditional blocks single chunk without body complex condition 1`] = `
Array [
  Object {
    "blockEnd": 28,
    "blockStart": 0,
    "bodyStart": 27,
    "chunkEnd": 0,
    "chunkStart": 0,
    "condition": Object {
      "prop": "b",
      "rest": " === [1, 2].join(\\" \\")",
    },
  },
]
`;

exports[`with conditional blocks single chunk without body multiple blocks 1`] = `
Array [
  Object {
    "blockEnd": 16,
    "blockStart": 9,
    "bodyStart": 15,
    "chunkEnd": 0,
    "chunkStart": 0,
    "condition": Object {
      "prop": "a",
      "rest": "",
    },
  },
  Object {
    "blockEnd": 39,
    "blockStart": 26,
    "bodyStart": 38,
    "chunkEnd": 0,
    "chunkStart": 0,
    "condition": Object {
      "prop": "a",
      "rest": " === b",
    },
  },
]
`;

exports[`with conditional blocks single chunk without body prop with operations 1`] = `
Array [
  Object {
    "blockEnd": 11,
    "blockStart": 0,
    "bodyStart": 10,
    "chunkEnd": 0,
    "chunkStart": 0,
    "condition": Object {
      "prop": "a",
      "rest": " + 1",
    },
  },
]
`;

exports[`with conditional blocks single chunk without body simple prop 1`] = `
Array [
  Object {
    "blockEnd": 7,
    "blockStart": 0,
    "bodyStart": 6,
    "chunkEnd": 0,
    "chunkStart": 0,
    "condition": Object {
      "prop": "a",
      "rest": "",
    },
  },
]
`;

exports[`with no conditional blocks incorrect conditions multiple chunks correct but in multiple chunks 1`] = `Array []`;

exports[`with no conditional blocks incorrect conditions multiple chunks incorrect keyword 1`] = `Array []`;

exports[`with no conditional blocks incorrect conditions multiple chunks keyword in multiple chunks 1`] = `Array []`;

exports[`with no conditional blocks incorrect conditions multiple chunks no brace 1`] = `Array []`;

exports[`with no conditional blocks incorrect conditions multiple chunks no matching brace 1`] = `Array []`;

exports[`with no conditional blocks incorrect conditions multiple chunks operator before identifier 1`] = `Array []`;

exports[`with no conditional blocks incorrect conditions multiple chunks without identifier 1`] = `Array []`;

exports[`with no conditional blocks incorrect conditions single chunk incorrect keyword 1`] = `Array []`;

exports[`with no conditional blocks incorrect conditions single chunk no brace 1`] = `Array []`;

exports[`with no conditional blocks incorrect conditions single chunk no condition body 1`] = `Array []`;

exports[`with no conditional blocks incorrect conditions single chunk no matching brace 1`] = `Array []`;

exports[`with no conditional blocks incorrect conditions single chunk operator before identifier 1`] = `Array []`;

exports[`with no conditional blocks incorrect conditions single chunk without identifier 1`] = `Array []`;

exports[`with no conditional blocks no conditions empty chunks 1`] = `Array []`;

exports[`with no conditional blocks no conditions multiple chunks 1`] = `Array []`;

exports[`with no conditional blocks no conditions no chunks 1`] = `Array []`;

exports[`with no conditional blocks no conditions single chunk 1`] = `Array []`;
